allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,244 @@
+        head_scores = self._head_scorer(contextualized_emb) # [b, text_len, 1]
+        head_offsets = util.get_indices(self._max_span_width, is_cuda).view(1, 1, -1) # [1, 1, max_span_width]
+        head_mask = (head_offsets <= span_size).float() # [b, num_spans, max_span_width]
+        raw_head_idx = span_ends - head_offsets # [b, num_spans, max_span_width]
+        head_mask = head_mask * (raw_head_idx >= 0).float() # [b, num_spans, max_span_width]
+        head_idx = F.relu(raw_head_idx.float()).long() # [b, num_spans, max_span_width]
+        flat_head_idx = util.flatten_batched_indices(head_idx,
+                                                     text_len) # [b * num_spans * max_span_width]
+        span_text_emb = util.batched_index_select(text_emb,
+                                                  head_idx,
+                                                  flat_head_idx) # [b, num_spans, max_span_width, emb]
+        span_head_scores = util.batched_index_select(head_scores,
+                                                     head_idx,
+                                                     flat_head_idx).squeeze(-1) # [b, num_spans, max_span_width]
+        span_head_scores += head_mask.float().log() # [b, num_spans, max_span_width]
+        flat_span_head_scores = span_head_scores.view(-1, self._max_span_width) # [b * num_spans, max_span_width]
+        flat_span_head_weights = F.softmax(flat_span_head_scores) # [b * num_spans, max_span_width]
+        flat_span_head_weights = flat_span_head_weights.unsqueeze(1) # [b * num_spans, 1, max_span_width]
+        flat_span_text_emb = span_text_emb.view(-1,
+                                                self._max_span_width,
+                                                span_text_emb.size(-1)) # [b * num_spans, max_span_width, emb]
+        flat_attended_text_emb = flat_span_head_weights.bmm(flat_span_text_emb) # [b * num_spans, 1, emb]
+        attended_text_emb = flat_attended_text_emb.view(batch_size, num_spans, -1) # [b, num_spans, emb]
+        span_emb = torch.cat([start_emb, end_emb, span_size_emb, attended_text_emb], -1) # [b, num_spans, emb]
+
+        # Compute mention scores.
+        mention_scores = self._mention_scorer(self._mention_feedforward(span_emb)) # [b, num_spans, 1]
+        mention_scores += span_mask.log() # [b, num_spans, 1]
+
+        # Prune based on mention scores.
+        k = int(math.floor(self._spans_per_word * text_len)) # []
+        _, top_span_idx = mention_scores.topk(k, 1) # [b, k, 1]
+        top_span_idx, _ = torch.sort(top_span_idx, 1) # [b, k, 1]
+        top_span_idx = top_span_idx.squeeze(-1) # [b, k]
+        flat_top_span_idx = util.flatten_batched_indices(top_span_idx, num_spans) # [b * k]
+
+        # Select tensors relating to the top spans.
+        top_span_mask = util.batched_index_select(span_mask, top_span_idx, flat_top_span_idx) # [b, k, 1]
```

Comment:
Is this assuming you could pick something in the top k that was masked?

Reply:
It could, but it would also conveniently end up masking the overall score, since the mention score is a factor in the antecedent score :)